vmm/src/serial_manager.rs

```diff
@@ -266,11 +310,63 @@ impl SerialManager {
                                 }
+                                EpollDispatch::Socket => {
+                                    // New connection request arrived.
+                                    // Shutdown the previous connection, if any
+                                    if let Some(previous_reader) = reader {
+                                        previous_reader
+                                            .shutdown(Shutdown::Both)
+                                            .map_err(Error::AcceptConnection)?;
+                                    }
+                                    // Events on the listening socket will be connection requests.
+                                    // Accept them, create a reader and a writer.
+                                    let (unix_stream, _) =
+                                        listener.accept().map_err(Error::AcceptConnection)?;
+                                    let writer =
+                                        unix_stream.try_clone().map_err(Error::CloneUnixStream)?;
+                                    reader = Some(
+                                        unix_stream.try_clone().map_err(Error::CloneUnixStream)?,
+                                    );
+
+                                    epoll::ctl(
+                                        epoll_fd,
+                                        epoll::ControlOptions::EPOLL_CTL_ADD,
+                                        unix_stream.into_raw_fd(),
+                                        epoll::Event::new(
+                                            epoll::Events::EPOLLIN,
+                                            EpollDispatch::File as u64,
+                                        ),
+                                    )
+                                    .map_err(Error::Epoll)?;
+                                    serial.lock().unwrap().set_out(Box::new(writer));
+                                }
                                 EpollDispatch::File => {
                                     if event.events & libc::EPOLLIN as u32 != 0 {
                                         let mut input = [0u8; 64];
-                                        let count =
-                                            in_file.read(&mut input).map_err(Error::ReadInput)?;
+                                        let count = match mode {
+                                            ConsoleOutputMode::Socket => {
+                                                if let Some(mut serial_reader) = reader.as_ref() {
+                                                    let count = serial_reader
+                                                        .read(&mut input)
+                                                        .map_err(Error::ReadInput)?;
+                                                    if count == 0 {
+                                                        warn!("Socket for serial console closed!");
```

Comment:
Seems resonable for the client to close the socket connection - I would prefer...

`info!("Remote end closed serial socket");`